src/main/java/org/bitsofinfo/hazelcast/discovery/docker/swarm/SwarmDiscoveryUtil.java

```diff
@@ -373,145 +375,168 @@ public void setHazelcastPeerPort(Integer hazelcastPeerPort) {
+					}
 				}
 			}
 		}
-	  }
-	  return false;
+		return false;
 	}
-	
+
 	/**
 	 * Discover containers on the relevant networks that match the given
-	 * service criteria. 
-	 * 
+	 * service criteria, using additionally a NullServiceFilter
+	 *
 	 * @param docker
 	 * @param relevantNetIds2Networks
 	 * @param criteria
 	 * @return set of DiscoveredContainer instances
 	 * @throws Exception
 	 */
+	private Set<DiscoveredContainer> discoverContainersViaCriteria(DockerClient docker,
+																   Map<String,Network> relevantNetIds2Networks,
+																   Service.Criteria criteria) throws Exception {
+		// no ServiceFilter provided, so use one with no constraints
+		return discoverContainersViaCriteria(docker, relevantNetIds2Networks, criteria, NullServiceFilter.getInstance());
+	}
+
+	/**
+	 * Discover containers on the relevant networks that match the given
+	 * service criteria and service filter
+	 * 
+	 * @param docker
+	 * @param relevantNetIds2Networks
+	 * @param criteria to be passed as a parameter to the /services request
+	 * @param serviceFilter additional criteria to be applied to Services returned from the /services request
+	 * @return set of DiscoveredContainer instances
+	 * @throws Exception
+	 */
 	private Set<DiscoveredContainer> discoverContainersViaCriteria(DockerClient docker, 
 			Map<String,Network> relevantNetIds2Networks, 
-			Service.Criteria criteria) throws Exception {
+			Service.Criteria criteria,
+			ServiceFilter serviceFilter) throws Exception {
 
 		Set<DiscoveredContainer> discoveredContainers = new HashSet<DiscoveredContainer>();
 
 		// find all relevant services given the criteria....
 		List<Service> services = docker.listServices(criteria);
-		
+
 		if (services == null) {
-			logger.info("SwarmDiscoveryUtil["+this.context+"] No service match for given criteria, docker.listServices(criteria) returned null.");
+			logger.fine("SwarmDiscoveryUtil[" + this.context + "] No service match for given criteria, docker.listServices(criteria) returned null.");
 			return discoveredContainers;
 		}
- 
+
 		if (services.size() == 0) {
-			logger.info("SwarmDiscoveryUtil["+this.context+"] No service match for given criteria, docker.listServices(criteria) returned 0");
+			logger.fine("SwarmDiscoveryUtil[" + this.context + "] No service match for given criteria, docker.listServices(criteria) returned 0");
 			return discoveredContainers;
 		}
-			
-		logger.info("SwarmDiscoveryUtil["+this.context+"] Number of services matching given criteria = " + services.size());
+
+		logger.fine("SwarmDiscoveryUtil[" + this.context + "] Number of services matching given criteria = " + services.size());
 
 		for (Service service : services) {
 
-			logger.info("SwarmDiscoveryUtil["+this.context+"] Processing service with name=" + service.spec().name());
+			if(serviceFilter.reject(service)) {
+				logger.fine("service with name " + service.spec().name() + " rejected by filter " + serviceFilter);
+			}
+			else {
```

Comment:
get rid of the huge if/else block wrap here, just `continue`

Reply:
Done.